File: Greedy-Algorithms/CarFueling.py
# python3
import sys


def compute_min_refills(distance, tank, stops):
    # write your code here
    stops.insert(0,0)
    stops.append(distance)
    stops_count=0
    last_stop=0
    while last_stop<len(stops)-2:
        capacity=tank
        last_refuel=last_stop
        while last_stop<len(stops)-1 and stops[last_stop+1]-stops[last_refuel]<capacity:
            last_stop+=1
        if last_stop==last_refuel:
            return -1
        if last_stop<len(stops)-1:
            stops_count+=1
    return stops_count

if __name__ == '__main__':
    d=int(input())
    m=int(input())
    n=int(input())
    stops=[int(stop) for stop in input().split()]
    print(compute_min_refills(d, m, stops))

File: Greedy-Algorithms/src/com/Algorithms/CarFueling.java
package com.Algorithms;

public class CarFueling {
    private static int getNumberStops(int full_capacity,int[] stops){
        int stops_count=0;
        int last_refuel=0;
        int last_stop=0;
        while(last_stop<stops.length-2){
            int capacity=full_capacity;
            last_refuel=last_stop;
            while(last_stop<stops.length-1 && stops[last_stop+1]-stops[last_refuel]<capacity){
                last_stop++;
            }
            if(last_stop==last_refuel){
                return -1;
            }
            if(last_stop<stops.length-1){
                stops_count++;
            }
        }
        return stops_count;
    }

    public static void testRefueling(){
        int distance=10;
        int full_capacity=3;
        int[] stops={0,1,2,5,9,distance};
        int result=getNumberStops(full_capacity,stops);
        System.out.println(result);
    }
}
